Implement Controller for growing persistent volumes

gnufied opened the pull request:
This PR implements API and controller plane changes necessary for doing controller side resize.

xref :  https://github.com/kubernetes/community/pull/657 

Also xref https://github.com/kubernetes/features/issues/284 

```
Add alpha support for allowing users to grow persistent volumes. Currently we only support volume types that just require control plane resize (such as glusterfs) and don't need separate file system resize. 
```

gnufied commented:
/retest

wongma7 commented:
it's a golint error, should add pkg/controller/volume/expand to hack/.golint_failures

gnufied commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1056,3 +1063,18 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+var _ volume.Expander = &glusterfsDiskExpander{}
+
+func (e *glusterfsDiskExpander) ExpandVolumeDevice(newSize resource.Quantity, oldSize resource.Quantity) error {
+	glog.Infof("********** Calling expand volume device **********")
```
@humblec do you think you will have time to implement the change needed here for the glusterfs volume type? I have already verified that, this function gets called when a glusterfs volume is resized. We just need to implement correct behaviour here. let me know.

humblec replied:
@gnufied Sure.. !! I am on it.

humblec replied:
https://github.com/humblec/kubernetes/commit/ecde2183594ec520bc03f7ded04bcd6c62f36766 Done :) Thanks!

gnufied commented:
/assign @jsafrane @rootfs

rootfs commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

rootfs commented on pkg/controller/volume/expand/cache/volume_resize_map_test.go:
```diff
@@ -0,0 +1,93 @@
+	}
+}
+
+// GetTestVolumeSpec returns a test volume spec
```
Get -> get

rootfs commented on pkg/controller/volume/expand/cache/volume_resize_map_test.go:
```diff
@@ -0,0 +1,93 @@
+				},
+			},
+		},
+		PersistentVolume: &v1.PersistentVolume{
```
a volume.Spec with 2 volumes is confusing.

gnufied replied:
you are correct. I copied this from validation_tests, we don't need both.

rootfs commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+	expc.resizeMap.AddPvcUpdate(newPvc, oldPvc, volumeSpec)
+}
+
+func CreateVolumeSpec(
```
is `CreateVolumeSpec` used outside of the package or `createVolumeSpec` would just work?

gnufied replied:
`createVolumeSpec` should work. Oversight on my part will fix.

rootfs commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+
+	claimClone, ok := clone.(*v1.PersistentVolumeClaim)
+
+	if !ok {
```
claimClone == nil || !ok

rootfs commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+func (expc *expandController) GetHostIP() (net.IP, error) {
+	return nil, fmt.Errorf("GetHostIP() not supported by expand controller's VolumeHost implementation")
```
GetHostIP not supported

rootfs commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+func (expc *expandController) GetNodeLabels() (map[string]string, error) {
+	return nil, fmt.Errorf("GetNodeLabels() unsupported in expandController")
```
ditto

jsafrane commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -337,6 +337,7 @@ func NewControllerInitializers() map[string]InitFunc {
 	controllers["route"] = startRouteController
 	controllers["persistentvolume-binder"] = startPersistentVolumeBinderController
 	controllers["attachdetach"] = startAttachDetachController
+	controllers["volume-expand"] = startVolumeExpandController
```
persistentvolume-expander?

jsafrane commented on pkg/api/types.go:
```diff
@@ -543,6 +543,31 @@ type PersistentVolumeClaimSpec struct {
+// These are valid conditions of Pvc
+const (
+	// An user trigger resize of pvc has been started
+	PvcResizeStarted PvcConditionType = "ResizeStarted"
```
api-conventions.md says: *Do not use abbreviations in the API, except where they are extremely commonly used, such as "id", "args", or "stdin".*

I am afraid that full PersistentVolumeClaimResizeStarted is necessary (+all other PVC abbreviations in this file should be expanded).

smarterclayton replied:
Initialisms in Go type fields must be capitalized `PVCResizeStarted` not `PvcResizeStarted`.  Another example: `urlField` and `URLField` are correct, not `UrlField`.

smarterclayton replied:
But the general comment holds, please expand all of these to be `PersistentVolumeClaimXXX`

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+	_, updateErr := resizeMap.kubeClient.CoreV1().PersistentVolumes().Update(pvClone)
+
+	if updateErr != nil {
+		glog.V(4).Infof("Erro updating pv %q with error : %v", pvClone.Name, updateErr)
```
Erro -> Error

jsafrane commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+	expc.resizeMap = cache.NewVolumeResizeMap(expc.kubeClient)
+
+	pvcInformer.Informer().AddEventHandler(kcache.ResourceEventHandlerFuncs{
+		UpdateFunc: expc.pvcUpdate,
```
Watching PVC update is not enough. You need to react to PVCs that were modified when the controller was down. I.e. AddFunc + check that PVC.Spec.Capacity > PVC.Status.Capacity and then check the bound PV.Spec.Capacity.

jsafrane replied:
Reaction to Delete event might be useful too, as `resizeMap` should be updated when user deletes the PVC and thus is not interested in its resize.

gnufied replied:
We will add a separate populator loop to handle pvc updates while controller was down (not in this PR)

jsafrane replied:
I still miss reaction to Added events. When the controller is down when user resizes a PVC, all you get is a PVC Added event with different pvc.status.capacity and spec.capacity. You need to react to this.

Which brings me back to periodical reconcilation of all PVCs and just checking their pvc.status.capacity vs. pvc.spec.capacity. You don't need caches / queues and it's bulletproff to missed events (e.g. when the controller is down).

gnufied replied:
Are you saying, we should replace event based handling of PVCs with periodically reconcilation of all PVCs? so basically - replace current event based design with one based on poll. Will that be *only* way of handling resizes or we still watch for events?

gnufied replied:
I am not opposed to periodic reconcilation of all PVCs via polling with some sort of populator loop. I will add that in a follow up PR. It is mentioned in design doc too. Such a populator loop will run less frequently though, to avoid causing high cpu usage. 

@jsafrane you seem to be implying that, if we used periodic polling of all PVCs as *only* way of finding out PVCs which needed resizing then we won't need to cache/queue the PVCs.  While that is correct, in a large cluster with 10K PVCs or more, we will causing quite a bit of CPU usage. On other hand -  if we keep polling interval too long, then there could be significant delay in time user submits resize request and when resize requests get processed.

wongma7 replied:
I think he means two things:
- we need `AddFunc` to handle Add events, not just `UpdateFunc`
- we should do periodic reconciliation of *all* PVCs to handle missed events. there is no need for a complicated cache in this case: AddFunc + UpdateFunc + periodic reconciliation will take care of everything. i.e. instead of having AddFunc/UpdateFunc add to a cache which is then acted upon, simply act as soon as AddFunc/UpdateFunc occurs.

gnufied replied:
Added a separate pvc_populator loop to handle case of missed events because of controller crash.

jsafrane commented on pkg/volume/volume.go:
```diff
@@ -193,6 +193,13 @@ type BulkVolumeVerifier interface {
 
+type Expander interface {
+	// ExpandVolume expands the volume device via cloudprovider api call
+	ExpandVolumeDevice(newSize resource.Quantity, oldSize resource.Quantity) error
```
I would return also the actual new size of the volume. Some storage backends might round the value to the nearest GiB and you should know the actual value.

gnufied replied:
ack

jsafrane replied:
still missing return of actual new size of the volume

jsafrane commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+	syncResize SyncVolumeResize
+
+	// Operation executor
+	opExecutor operationexecutor.OperationExecutor
```
I thought that OperationExecutor is attach/detach specific. It may be not, but is it worth using here? Won't simple GoRoutineMap do the same job?

gnufied replied:
As discussed one of the other benefits of using common operation executor is, we get recently introduced volume metrics for free. Using operation executor might be more desirable in that sense.

jsafrane replied:
ok

jsafrane commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+	// PVC that needs to be resized
+	PVC *v1.PersistentVolumeClaim
+	// Volume spec object that can be used from volume plugins
+	VolumeSpec *volume.Spec
```
Why VolumeSpec? Do you plan to resize inline volumes too?

gnufied replied:
This is not inline volume spec. This volume spec is one defined in volume package which plugins know how to consume. Typically, most plugin calls require `volume.Spec` starting with `FindVolumePluginBySpec` .

jsafrane replied:
To me it would look better if you store PV here (because that's what you have from informers) and create VolumeSpec just before calling volume plugin.

gnufied replied:
done

jsafrane replied:
I still see `VolumeSpec *volume.Spec`

gnufied replied:
hmm, why though?

jsafrane reviewed (changes requested):
Current design is quite fragile. It does not count with PV/PVCs being updated/deleted by user or other controller when resize is in progress.

I would rework `syncResize` loop to list all bound PVCs, check their PVC.Spec.Capacity against PVC.Status.Capacity (+ maybe also PV.Spec.Capacity) and start a new named operation when they don't match and require some work.

The operation would look at *current* PVC and PV (from informers) and do whatever needs to be done - set/clear conditions & resize PV (& set/clear conditions again). It would do nothing when everything is ready and just waiting for fs resize. On error, it would return error from the operation and it would be automatically re-enqueed to a goroutinemap with exponential backoff.

The operations needs to "succeed" when user deletes the PVC - there is nothing to resize any longer.

IMO, there is no need to keep `VolumeResizeMap` at all.

jsafrane commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to find PV %q in PV informer cache : %v", volumeName, err)
+	}
+
```
You must check that the returned PV is bound to the PVC, i.e. its claimRef points back to the PVC (incl. claimRef.uid). Attacker could forge a PVC1 that points to a PV that is bound to PVC2 and then resize PVC1. The PV would be resized, taking quota from the attacker, but giving the actual space to user of PVC2.

gnufied commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+
+// VolumeResizeMap defines an interface that serves as a cache for holding pending resizing requests
+type VolumeResizeMap interface {
+	AddPvcUpdate(newPvc *v1.PersistentVolumeClaim, oldPvc *v1.PersistentVolumeClaim, spec *volume.Spec)
```
ack

gnufied commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+// performing file system resize
+type PvcWithResizeRequest struct {
+	// PVC that needs to be resized
+	PVC *v1.PersistentVolumeClaim
```
Discussed this offline. We will handle this problem via, if updating pvc or PV fails, then we will fetch latest version of PVC object from informer and rather than adding stale version of PVC, we will add back the fresh version of PVC to the VolumeResizeMap.

This will ensure that, next iteration of sync loop - resize will be attempted again (and could be noop if underlying volume was already resized) and save objects again.

gnufied replied:
done. Also added a populator loop for taking care of missed events.

saad-ali commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -188,6 +189,21 @@ func startAttachDetachController(ctx ControllerContext) (bool, error) {
 }
 
+func startVolumeExpandController(ctx ControllerContext) (bool, error) {
+	expandController, expandControllerErr := expand.NewExpandController(
```
Check `if utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes) {` and don't start the controller if the feature is not enabled.

saad-ali commented on pkg/api/validation/validation.go:
```diff
@@ -1553,10 +1553,36 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
+		newSize := newPvc.Spec.Resources.Requests["storage"]
+
+		if newSize.Cmp(oldSize) < 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be lesser than previous value"))
```
nit: `field can not be lesser than previous value` -> `field can not be less than previous value.`

saad-ali commented on pkg/volume/plugins.go:
```diff
@@ -199,6 +199,13 @@ type AttachableVolumePlugin interface {
+// expanded
+type ExpandableVolumePlugin interface {
+	VolumePlugin
+	NewExpander(spec *Spec) (Expander, error)
```
Any reason not to add the `ExpandVolumeDevice(...)` and `RequiresFSResize(...)` methods directly here instead of burying them inside `Expander`? A lot of the code in `volumes.go` is unnecessary and we should clean that up. But for new code let's not add unnecessary cruft to match the existing code.

saad-ali commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1045,3 +1056,62 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+	}
+
+	// Find out delta size
+	expansionSize := int(newSize.Value() - oldSize.Value())
```
Is the `oldSize` guaranteed to be accurate? If not, your delta could be wrong.

gnufied replied:
`oldSize` is picked from `pvc.Status.Capacity`  and since there can be only one resize operation active at a time for a volume, it is guaranteed to be correct. If we were to pick it from `pvc.Spec.Resources.Requests["Storage"]` or something - it would be inaccurate, but reading from `Status` should give us correct result.

humblec replied:
Yep. thats what my understanding too

saad-ali commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1045,3 +1056,62 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+	// Fetch the volume for expansion.
+	volumeID := dstrings.TrimPrefix(volumeName, volPrefix)
+
+	//Get details of SC.
```
Is fetching storageClass here required? It looks like it is only used for logging. In that case, if you fail to fetch SC, should that maybe be a warning instead of a terminating error?

gnufied replied:
It is also used for extracting heketi url etc from storageClass.

humblec replied:
its required @saad-ali. We need the REST url from SC of the PV at time of expansion.

saad-ali commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+		if rc.opsExecutor.IsOperationPending(uniqueVolumeKey, "") {
+			glog.V(10).Infof("Operation for PVC %v is already pending", pvcWithResizeRequest.UniquePvcKey())
+			continue
+		}
```
```GO
			glog.V(5).Infof("Starting opsExecutor.ExpandVolume...", volumeInfo...)
```

saad-ali commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+			glog.V(10).Infof("Operation for PVC %v is already pending", pvcWithResizeRequest.UniquePvcKey())
+			continue
+		}
+		growFuncError := rc.opsExecutor.ExpandVolume(pvcWithResizeRequest, rc.resizeMap)
```
```GO
			if err == nil {
				glog.Infof("opsExecutor.ExpandVolume started...", volumeInfo...)
			}
```

saad-ali commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+			continue
+		}
+		growFuncError := rc.opsExecutor.ExpandVolume(pvcWithResizeRequest, rc.resizeMap)
+		if growFuncError != nil {
```
Handle exponential back off errors:
```GO
			if err != nil && !exponentialbackoff.IsExponentialBackoff(err) {
				// Ignore exponentialbackoff.IsExponentialBackoff errors, they are expected.
				// Log all other errors.
			    glog.Errorf("Error growing pvc with %v", growFuncError)
			}
```

saad-ali commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+		}
+		growFuncError := rc.opsExecutor.ExpandVolume(pvcWithResizeRequest, rc.resizeMap)
+		if growFuncError != nil {
+			glog.Errorf("Error growing pvc with %v", growFuncError)
```
Add information about the volume, consider using `GenerateErrorDetailed(...)` in operation executor.

gnufied replied:
I don't think we can use `GenerateErrorDetailed` btw, because depends on stuff from `VolumeToAttach` struct, which is not available here.

saad-ali commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,69 @@
+	// Resize PVCs that require resize
+	for _, pvcWithResizeRequest := range rc.resizeMap.GetPvcsWithResizeRequest() {
+		uniqueVolumeKey := v1.UniqueVolumeName(pvcWithResizeRequest.UniquePvcKey())
+		if rc.opsExecutor.IsOperationPending(uniqueVolumeKey, "") {
```
Should there be some coordination to prevent resize during attach/detach operations?

gnufied replied:
We do have to prevent force detach if resize is in progress, but that change will be done as part of file system resize PR (will open todayish).

saad-ali commented on pkg/api/validation/validation.go:
```diff
@@ -1553,10 +1553,36 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
+
+		if !apiequality.Semantic.DeepEqual(newPvc.Spec.StorageClassName, oldPvc.Spec.StorageClassName) {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "storageClassName"), "field is immutable after creation"))
+		}
```
Should you validate that the volume is actually bound?

gnufied replied:
ack. I will have to add bound volume check here and perhaps in controller/sync_volume_resize too

saad-ali commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,285 @@
+
+// PvcWithResizeRequest struct defines data structure that stores state needed for
+// performing file system resize
+type PvcWithResizeRequest struct {
```
You are blurring functionality here. The purpose of the cache should be very crisp and clear: it stores in-memory state. It should not construct any requests, that should be the job of the reconciler, aka `sync_volume_resize.go`. In fact, the `volumeResizeMap struct` above should not even have a `kubeClient clientset.Interface` at all.

gnufied replied:
I agree that `VolumeResizeMap` is more than just a cache of PVCs which needs to be resized, it also contains interface of functions which update the PVC after success/failure. 

There is certain simplicity to it, which makes testing easier. For example, if resize failed then the certain version of PVC must be added back to `VolumeResizeMap`. We can test that directly and it involves no goroutines and hence setting up the test scenario is simpler.  Every function call in `VolumeResizeMap` is direct and has no asynchronous behaviour. 

If we decouple updating PVC after resizing from `VolumeResizeMap` then - the reconciler will have to perform those updates. But it isn't as simple as that, because - sync_volume_resize calls volume operations through a goroutine (operation executor), it doesn't know about success or failure of particular operation, so in that case we will have to keep track of successfully resized PVCs separately.  So, we will have to introduce actual and desired state of world (or something along those lines).

But there is also inherent risk in keeping PVC objects in cache for too long (and in more than one place) - as jan pointed out, you may end up with stale version of PVC object and updating it may fail in etcd. 

The original version of code I wrote did use - actual & desired state of world but then I realized the design can be simplified if we just used one cache.

saad-ali commented on pkg/controller/volume/expand/BUILD:
```diff
@@ -0,0 +1,54 @@
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
```
Add a new OWNER file for the new controller. Populate it with me, you, and Jan as approvers.

saad-ali reviewed (changes requested):
Took a first pass

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -272,7 +303,23 @@ func (resizeMap *volumeResizeMap) addBackResizeRequest(pvcr *PvcWithResizeReques
 		return
 	}
-	resizeMap.pvcrs[pvcUniqueName] = pvcr
+	pvc, err := resizeMap.kubeClient.CoreV1().PersistentVolumeClaims(pvcr.PVC.Namespace).Get(pvcr.PVC.Name, metav1.GetOptions{})
```
Use shared informers please.

gnufied replied:
This object doesn't have informers. It can be supplied to it, if need be - but I am not sure if it should be a blocker. Doing one off, GET should be fine directly via kubeClient?

gnufied reviewed (commented):
@saad-ali @jsafrane PTAL. I have pushed following changes (since you probably last reviewed)

* Remove additional pvc.Status.Condition and make use of just one.
* Add an additional populator loop for taking care of missed events.
* Handle stale PVC object on update etc everywhere.
* Add more tests

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,303 @@
+	return nil
+}
+
+func (resizeMap *volumeResizeMap) MarkForFileSystemResize(pvcr *PvcWithResizeRequest) error {
```
this function should get actual new size of the volume, it may be rounded up.

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,303 @@
+	// If any part of this update fails, resize has to be tried again
+	emptyCondition := []v1.PersistentVolumeClaimCondition{}
+
+	_, err := util.UpdatePVCCondition(pvcr.PVC, emptyCondition, resizeMap.kubeClient)
```
IMO we should keep any ResizeStarted condition there, the controller is still trying to resize the PVC.

gnufied replied:
We do set it back when we actually attempt resize in `sync_resize_map`. Should be enough may be?

jsafrane replied:
I'd leave the condition there. Constant adding + removing the condition looks weird to me. The condition should tell user that resizing is in progress on the PVC and has not finished yet. And it has not finished, the controller is just taking a break for a while.

gnufied replied:
done

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,303 @@
+		// rather than pvc.Status.Size is because if a resize is already in progress
+		// then pv size would reflect new value and if pv size is more than pvc
+		// size then no resize will be necessary.
+		if pvSize.Cmp(pvcSize) < 0 {
```
You should compare also pvc.status.capacity with pvc.spec.capacity. For example, when the controller updates pv.spec.capacity with new value and crashes just before saving pvc.status.capacity then there is no path that would make sure that something updates the status.

Again, the best would be check *only* pvc.spec vs pvc.status and the operation itself would check what needs to be done - update pvc status conditions, resize PV, update pv.status, update pvc.status. There is some overhead when the controller is waiting for FS resize, but that's IMO it's negligible.

gnufied replied:
I have thought long and hard about it and comparing `pvc.spec.capacity` and `pvc.status.capacity` here - will result in double resizing of every volume type that requires file system resize.  IMO it is unnecessary churn and optimizing for the worst 2% of case (namely controller crash). 

Is there a way, we can handle this and avoid double resizing of same volume?  I am thinking that, rather than solving this ourselves - may be we should provide admin mechanism to sort it out. namely:

1. We can log an event if we notice that a pvc's status and spec capacity don't match and is in `ResizeStarted` state for too long..
2. If a pvc's status and spec doesn't match and is not in `ResizeInProgress`, then it gets added to resize map.

So basically what we are saying is, admin can manually edit the PVC and drop `ResizeInProgress` field from pvc.Status and that would make populator pick it up again.

gnufied replied:
There is another problem with solely relying on `pvc.Spec.Capacity` and `pvc.Status.Capacity`. When you create a new PVC and when it gets dynamically provisioned - I am seeing a state where `pvc.Spec.Capacity` has user requested size but `pvc.Status.Capacity` is zero.  It is temporary and it does seem to get fixed in next update. But relying only on `pvc.Spec.Capacity` and `pvc.Status.Capacity` would result in expand controller trying to resize such a volume.

jsafrane replied:
> When you create a new PVC and when it gets dynamically provisioned - I am seeing a state where pvc.Spec.Capacity has user requested size but pvc.Status.Capacity is zero.

Is the PVC bound at that time? It's error if it's reported as Bound and status.capacity does not reflect the PV size.

jsafrane replied:
> will result in double resizing of every volume type that requires file system resize

```golang

func resizeOperation(pv, pvc) {
    if pvc.spec.Capacity <= pvc.status.Capacity {
        return
    }

    if !hasResizingCondition(pvc.status) {
        // Add the condition if it's not there yet
        addResizingCondition(pvc)
    }

    if pv.spec.capacity < pvc.spec.capacity {
        // PV has not been resized yet, resize it
        newSize = plugin.ExpandVolumeDevice(pvc.spec.capacity)
        saveNewPVSize(pv, newSize)
    }

    if needFSResize() {
        // waiting for FS resize, it will update pvc.status.capacity and remove the condition in the end
        return
    } else {
        saveNewPVCSizeAndRemoveResizingCondition(pvc, newSize)
    }
}
```

This should be idempotent and can be interrupted at any time without any loss. It extends the volume twice only when it crashes between `plugin.ExpandVolumeDevice` and `saveNewPVSize`, which is IMO rare enough. If anything returns error, the operation can be restarted and it will do only the necessary changes.

When a PV waits for a FS resize, this operation just checks that a condition is set, compares some sizes and returns.

gnufied replied:
ack done

gnufied commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,303 @@
+
+// VerifyAndSoftAddPVC adds PVC for resizing only if there is no
+// resize request pending and if PV size doesn't satisfy pvc size
+func (resizeMap *volumeResizeMap) VerifyAndSoftAddPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
```
see comment below

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,303 @@
+	}
+	pvc, err := resizeMap.kubeClient.CoreV1().PersistentVolumeClaims(pvcr.PVC.Namespace).Get(pvcr.PVC.Name, metav1.GetOptions{})
+	if err != nil {
+		glog.Errorf("Error fetching pvc %v/%v for resize with error : %v", pvc.Namespace, pvc.Name, err)
```
This will happen when someone deletes the PVC during resize - say `plugin.ExpandVolumeDevice` fails and the request should be enqueued again and it ends up here. A lower log level would be nice + some comment that  this is expected behavior. It's not obvious that seemingly endless loop of `ExpandVolumeDevice` error + `addBackResizeRequest` for deleted volumes stops here.

gnufied replied:
ack done.

jsafrane commented on pkg/controller/volume/expand/pvc_populator.go:
```diff
@@ -0,0 +1,89 @@
+
+	for _, pvc := range pvcs {
+		// We need to only process bound pvcs
+		if pvc.Status.Phase == v1.ClaimBound {
```
IMO, this check already done in `pvcUpdate`, it's going to have periodic sync every 5-10 minutes.
In other words, there should be a function `checkPVCForResize` that would check the bound/unbound status, get the PV and eventually call `AddPVCUpdate` so these checks are on one place.

jsafrane replied:
edited, sorry... Trying to do a quick review.

jsafrane commented on pkg/controller/volume/expand/sync_volume_resize.go:
```diff
@@ -0,0 +1,101 @@
+			glog.V(10).Infof("Operation for PVC %v is already pending", pvcWithResizeRequest.QualifiedName())
+			continue
+		}
+		glog.V(5).Infof(" Starting opsExecutor.ExpandVolume for volume %s", pvcWithResizeRequest.QualifiedName())
```
extra space in `" Starting`

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -726,6 +729,60 @@ func (og *operationGenerator) verifyVolumeIsSafeToDetach(
+				pvcWithResizeRequest.CurrentSize)
+
+			if expandErr != nil {
+				glog.Errorf("Error expanding volume through cloudprovider for volume %q : %v", pvcWithResizeRequest.QualifiedName(), expandErr)
```
not all volumes use CloudProvider, something like `error expanding volume %q: %v`. And if you want, you can log `volumePlugin.GetPluginName`

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,256 @@
+
+// VolumeResizeMap defines an interface that serves as a cache for holding pending resizing requests
+type VolumeResizeMap interface {
+	AddPVCUpdate(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume)
```
these needs some documentation (+ consistent empty lines - the one below looks weird).

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+// UpdatePVSize updates just pv size after cloudprovider resizing is successful
+func (resizeMap *volumeResizeMap) UpdatePVSize(pvcr *PVCWithResizeRequest, newSize resource.Quantity) error {
```
this function  must call `addBackResizeRequest` on error so the operation is re-tried.

jsafrane replied:
this comment is no longer valide, addBackResizeRequest  was removed

wongma7 commented on plugin/pkg/admission/persistentvolume/resize/admission.go:
```diff
@@ -0,0 +1,144 @@
+	return nil
+}
+
+// allowResize checks whether the pvc is dynamically provisioned and if the storageclass allows expansion
```
non-empty class doesn't necessarily mean the pvc had its pv dynamically provisioned for it, the pv could have been manually created and added to the class.

So I would edit this comment, I think we only care if the storageclass allows expansion, not if the pv was dynamically provisioned

wongma7 replied:
nvm, the design does say the pv must have been dynamically provisioned.

this means this check has to be made stronger, it has to get the pv to check if it was dynamically provisioned?

jsafrane replied:
IMO we should not block resizes of statically provisioned PVs if the plugin + storage class allows it

wongma7 replied:
+1, let's just edit this and https://github.com/kubernetes/kubernetes/pull/49727#discussion_r136439045 to clarify the actual behavior

jsafrane commented on pkg/apis/storage/types.go:
```diff
@@ -59,6 +59,10 @@ type StorageClass struct {
+
+	// AllowVolumeExpand shows whether the storage class allow volume expand
+	// +optional
+	AllowVolumeExpand *bool
```
Missing documentation of expected behavior when the field is missing/nil - is resize allowed of forbidden?

smarterclayton replied:
`AllowVolumeExpansion` or `AllowExpansion` seem more accurate.  Was there a discussion on the naming in the proposal?

smarterclayton replied:
this field should be "AllowVolumeExpansion"

jsafrane commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,201 @@
+// AddPVCUpdate adds pvc for resizing
+func (resizeMap *volumeResizeMap) AddPVCUpdate(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+	if pvc.Namespace != pv.Spec.ClaimRef.Namespace || pvc.Name != pv.Spec.ClaimRef.Name {
+		glog.Errorf("Persistent Volume is not mapped to PVC being updated : %s", util.ClaimToClaimKey(pvc))
```
I don't think that Error is the right level. This is going to be logged every 2 minutes for all pre-bound PVCs... Maybe V(2)?

gnufied replied:
ack, I am going to put this to V(4)

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -22,11 +22,12 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/golang/glog"
```
please restore original order

jsafrane commented on plugin/pkg/admission/persistentvolume/resize/admission.go:
```diff
@@ -0,0 +1,144 @@
+		return nil
+	}
+
+	// Only dynamically provisioned pvc/pv can be resized
```
This comment is misleading, it's not about dynamic/static but about storageclass allows resize or not.

jsafrane commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -124,6 +124,14 @@ func init() {
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "expand-controller"},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("update").Groups(legacyGroup).Resources("persistentvolumes").RuleOrDie(),
+			rbac.NewRule("update").Groups(legacyGroup).Resources("persistentvolumeclaims/status").RuleOrDie(),
```
Gluster plugin requires reading storage classes, should it be included here?

jsafrane replied:
fixed (few lines below)

jsafrane commented:
There are some nits reported above, but in general this PR is almost lgtm. It's good time to hunt for approvals.

jsafrane commented:
/release-note

jsafrane commented:
/lgtm

jsafrane commented:
I'd appreciate an e2e test. @humblec, what's the progress on e2e test for gluster provisioner? We could squeeze resize there easily.

smarterclayton commented on pkg/api/validation/validation.go:
```diff
@@ -1553,10 +1553,28 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
-	if !apiequality.Semantic.DeepEqual(newPvc.Spec, oldPvc.Spec) {
-		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "field is immutable after creation"))
+
+	oldSize := oldPvc.Spec.Resources.Requests["storage"]
```
Old and new can be nil?

smarterclayton replied:
Nm, Requests is not a pointer.

smarterclayton commented on pkg/api/validation/validation.go:
```diff
@@ -1553,10 +1553,28 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
+	newSize := newPvc.Spec.Resources.Requests["storage"]
+
+	if newSize.Cmp(oldSize) < 0 {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be lesser than previous value"))
```
`can not be less than the previous value`

gnufied replied:
fixed

smarterclayton commented on pkg/api/validation/validation.go:
```diff
@@ -1553,10 +1553,28 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be lesser than previous value"))
+	}
+
+	if !apiequality.Semantic.DeepEqual(newPvc.Spec.AccessModes, oldPvc.Spec.AccessModes) {
```
this pattern is dangerous because it can lead to accidental mutation (someone adds a new field and forgets to add it here).  The better path is to copy spec, check anything that is allowed to change, then copy the allowed to change field to the copy, then do a a deep equal.  Other validations do this, and it's the right thing to do.

gnufied replied:
fixed.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -630,6 +630,36 @@ type PersistentVolumeClaimSpec struct {
+	Status ConditionStatus  `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+	// Last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,3,opt,name=lastProbeTime"`
```
If you don't plan on setting LastProbeTime omit it from this struct.  We generally don't set it because it's expensive to calculate.

gnufied replied:
I do plan to keep it for now, because when we implement file system resize,  we want to report on progress of resize. We may have to introduce additional conditions at that point though. But for now, I am leaving it as it is.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -630,6 +630,36 @@ type PersistentVolumeClaimSpec struct {
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+	// Unique, one-word, CamelCase reason for the condition's last transition.
```
The description of this should look more like the Event Reason godoc:

```
	// Optional; this should be a short, machine understandable string that gives the reason
	// for this event being generated. For example, if the event is reporting that a container
	// can't start, the Reason might be "ImageNotFound".
```

Saying that it is CamelCase doesn't help a human understand what it is for.  You should describe in this description which conditions are normally set and also explain that other types are possible.

gnufied replied:
added more info there.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -642,6 +672,11 @@ type PersistentVolumeClaimStatus struct {
 	// Represents the actual resources of the underlying volume.
 	// +optional
 	Capacity ResourceList `json:"capacity,omitempty" protobuf:"bytes,3,rep,name=capacity,casttype=ResourceList,castkey=ResourceName"`
+	// Current service state of PVC
```
Please add more description here about what conditions are normally set and how an end user might interpret this.

gnufied replied:
done

smarterclayton commented on pkg/api/validation/validation.go:
```diff
@@ -1602,6 +1602,10 @@ func ValidatePersistentVolumeClaimStatusUpdate(newPvc, oldPvc *api.PersistentVol
 	if len(newPvc.Spec.AccessModes) == 0 {
 		allErrs = append(allErrs, field.Required(field.NewPath("Spec", "accessModes"), ""))
 	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes) && len(newPvc.Status.Conditions) > 0 {
```
I don't think your feature gate has to protect conditions.  Conditions are something any controller can set.  You can remove this check.

smarterclayton replied:
Keep the test if it's valid though.

liggitt replied:
the conditions field is new to this feature... it's a tad odd to add it and unconditionally enable it with only one current use

gnufied replied:
yeah, I have left this check in place for now.

smarterclayton commented on pkg/apis/storage/validation/validation.go:
```diff
@@ -108,3 +111,13 @@ func validateReclaimPolicy(reclaimPolicy *api.PersistentVolumeReclaimPolicy, fld
+func validateAllowVolumeExpand(allowExpand *bool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if allowExpand != nil && !utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes) {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "AllowVolumeExpand field is disabled by feature-gate ExpandPersistentVolumes"))
```
Field names, when referenced in user facing godoc or error messages, should always be in the form of their json equivalent.  In this case the field path already contains the name of the field (or should), in which case the message should be `expansion of volumes is disabled by the feature-gate ...`

gnufied replied:
If feature gate is disabled - it stops users from setting the field on StorageClass. So I have changed this to "field is disabled by feature-gate ExpandPersistentVolumes"

smarterclayton commented:
API is close, the field name needs to be clarified and the validation definitely needs to be made safe.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -124,6 +124,17 @@ func init() {
 		},
 	})
 	addControllerRole(rbac.ClusterRole{
```
cc @kubernetes/sig-auth-pr-reviews

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -124,6 +124,17 @@ func init() {
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "expand-controller"},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("update").Groups(legacyGroup).Resources("persistentvolumes").RuleOrDie(),
+			rbac.NewRule("update").Groups(legacyGroup).Resources("persistentvolumeclaims/status").RuleOrDie(),
```
we try to keep roles defined so that if the controller was run solo in a pod, the role would be sufficient for its needs. I'm assuming this makes use of list/watch on PVs, and if `update` is granted, `get` and `patch` are typically allowed as well.

gnufied replied:
I had to grant permissions for reading secrets, endpoints, services as well to expand-controller's client.  It should have enough permissions to run standalone now.

liggitt replied:
what client was it using to get secrets before? the attach/detach controller doesn't have secret access...

gnufied replied:
Attach/Detach controller doesn't handle glusterfs. The reason, we had to add secret here is because there was a change made for glusterfs Storageclass so that glusterfs credentials are no longer directly available in SC (you should remember this change!). 

So credentials for heketi are stored in another namespace and this client must be able to read them - to be able to resize glusterfs volumes.

Again to reiterate - since glusterfs is not a attachable volume type, attach/detach controller doesn't do anything with it.  Originally when I tested this PR, I didn't had rbac enabled cluster but other people who have contributed to this PR ( @wongma7 , @humblec) did. Now, in addition to them - I have tested with rbac enabled cluster as well and these permissions here are needed.

liggitt commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -188,6 +189,24 @@ func startAttachDetachController(ctx ControllerContext) (bool, error) {
+			ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+			ctx.InformerFactory.Core().V1().PersistentVolumes(),
+			ctx.Cloud,
+			ProbeExpandableVolumePlugins(ctx.Options.VolumeConfiguration))
```
does this need a peer to `GetDynamicPluginProber(ctx.Options.VolumeConfiguration)`?

gnufied replied:
We don't need `GetDynamicPluginProber` for this controller (at least for now), because currently only flex volumes are dynamically probed and they are not resizable.

liggitt replied:
> currently only flex volumes are dynamically probed and they are not resizable.

why not? keeping parity in our out-of-tree volume mechanisms is important

gnufied replied:
For resizing to work for Flex volumes, we will have to add expand/resize API calls to flex protocol. We are yet to be in a place where flex volumes can be called stable. Attach/Detach interface was added recently to flex and has issues that we are trying to workout - https://github.com/kubernetes/kubernetes/issues/44737. sig-storage is also working on making flex plugin deployments saner. 

In a nutshell, I would like to pursue flex resizing as separate problem. Several community members have expressed an interest in solving it. There is a separate feature request open for it- https://github.com/kubernetes/features/issues/304

gnufied replied:
But just to clarify current resizing design does not inhibit a future flex resize design. As long as, protocols/API calls laid out by this design are implemented by flex - we will have working resizing. So it all comes down to implementing volume plugin interface necessary for resizing.

liggitt replied:
understood. seems like that should tracked as part of graduating this to beta.

gnufied replied:
ack. I will make a note of this in original design proposal.

gnufied commented:
@smarterclayton addressed most of your concerns. PTAL.

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1556,10 +1556,31 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeCla
+
+		// lets make sure storage values are same.
+		if newPvc.Status.Phase == api.ClaimBound {
+			newPVCSpecCopy.Resources.Requests["storage"] = oldPvc.Spec.Resources.Requests["storage"]
```
can Requests be nil? if so, this will panic

gnufied replied:
It is not a pointer value and so I would think it would be initialized to its default value, which is a map.

liggitt replied:
default value of a map is nil, which panics if assigned to

gnufied replied:
fixed.

gnufied commented:
/test pull-kubernetes-kubemark-e2e-gce

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -124,6 +124,19 @@ func init() {
 		},
 	})
 	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "expand-controller"},
```
gate this on `utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes)`

gnufied replied:
Feature gating the controller policy - somehow disables controller access to Update etc, even though feature gate is enabled for every component including apiserver.  I have enabled it via `export FEATURE_GATES=blah` and then `./hack/local-up-cluster.sh` and verified that apiserver cli does contain feature gate.

gnufied replied:
As discussed offline, there appears to be a bug in the way policies are bootstapped. @liggitt cooked up a patch for me - https://gist.github.com/liggitt/8cbf0c907d06f4ca3c093073600f1ef7 which works.  But I am uncomfortable carrying a patch that I didn't write and understand. 

I think, we should leave this policy change  as it is, until the behaviour is fixed in master and then I am happy to feature gate this policy change.

liggitt replied:
>  I am uncomfortable carrying a patch that I didn't write and understand.

https://github.com/kubernetes/kubernetes/pull/51818 is open to build the roles on demand, which would make feature gates effective during startup reconcile.

> I think, we should leave this policy change as it is, until the behaviour is fixed in master and then I am happy to feature gate this policy change.

Merging this ungated automatically initializes these roles/bindings on start, which is not what we want.

liggitt commented on pkg/controller/volume/expand/util/util.go:
```diff
@@ -0,0 +1,67 @@
+		return nil, err
+	}
+	claimClone.Status.Conditions = pvcConditions
+	updatedClaim, updateErr := kubeClient.CoreV1().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
```
this should really patch to ensure version skewed clients don't stomp unknown fields

gnufied replied:
This isn't getting called from version skewed clients like kubelet yet. I will open a github issue for myself to move this over to patch before it does. let me know if that is fine.

liggitt commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,201 @@
+
+	claimClone.Status.Capacity[v1.ResourceStorage] = newSize
+	claimClone.Status.Conditions = pvcConditions
+	_, updateErr := resizeMap.kubeClient.CoreV1().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
```
patch

gnufied replied:
same as above.

liggitt commented on pkg/controller/volume/expand/util/util.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// ClonePVC returns a clone of PVC to be used while updating it
+func ClonePVC(oldPvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
```
not needed, call `oldPvc.DeepCopy()`

liggitt commented on pkg/controller/volume/expand/cache/volume_resize_map.go:
```diff
@@ -0,0 +1,201 @@
+	defer resizeMap.Unlock()
+
+	oldPv := pvcr.PersistentVolume
+	clone, err := scheme.Scheme.DeepCopy(oldPv)
```
`oldPv.DeepCopy()`

liggitt commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -0,0 +1,277 @@
+		return nil, fmt.Errorf("failed to find PV %q in PV informer cache with error : %v", volumeName, err)
+	}
+
+	clonedPvObject, err := scheme.Scheme.DeepCopy(pv)
```
`pv.DeepCopy()`

derekwaynecarr commented on pkg/quota/evaluator/core/persistent_volume_claims.go:
```diff
@@ -144,7 +144,7 @@ func (p *pvcEvaluator) GroupKind() schema.GroupKind {
 func (p *pvcEvaluator) Handles(a admission.Attributes) bool {
 	op := a.GetOperation()
-	if op == admission.Create {
+	if op == admission.Create || op == admission.Update {
```
open an issue for a follow-on to ensure that on update, we change the max(new,old) per resource.

gnufied replied:
yep done here - https://github.com/kubernetes/kubernetes/issues/51736

liggitt replied:
should only handle update if the feature gate is enabled... otherwise, you're opening PVs to negative quota spoofing on update

liggitt replied:
actually, this already interacts oddly with the uninitialized bit below

jsafrane replied:
added feature gate check

gnufied commented:
@liggitt I believe all open issues were addressed. cc @smarterclayton

childsb commented:
This PR also includes #51818 which is in queue, but not yet merged.  Once #51818 merges this may need rebase, but is approved/targeted for 1.8

jsafrane commented:
/lgtm

liggitt commented:
the end result of the RBAC changes LGTM, I haven't reviewed the rest.

the commits should be squashed down and made coherent... there's a lot of churn in the commit history that makes reviewing difficult. I'd expect something like:
* clean pick of https://github.com/kubernetes/kubernetes/pull/51818
* api, feature gate, validation changes
* controller changes
* generated files

smarterclayton commented:
Please make those changes as jordan requested, will check back in before too long.

gnufied commented:
@liggitt @smarterclayton done. PTAL

childsb commented:
/lgtm

smarterclayton commented on pkg/api/types.go:
```diff
@@ -547,6 +547,27 @@ type PersistentVolumeClaimSpec struct {
+// These are valid conditions of Pvc
+const (
+	// An user trigger resize of pvc has been started
+	PersistentVolumeClaimResizeStarted PersistentVolumeClaimConditionType = "ResizeStarted"
```
This can be a follow up but should be "Resizing", not "ResizingStarted".  A condition is level driven, not edge driven

gnufied replied:
You mean both Constant and its value should be - `PersistentVolumeClaimResizing` and `Resizing` ?

gnufied replied:
done.

smarterclayton commented:
Two API changes.  Please fix those and ping me for final review, but I will mark this as api approved so that we're covered for 1.8 code freeze.

/approve

With the understanding the two issues will be addressed before merge.

gnufied commented:
@smarterclayton both API changes have been done. PTAL.

childsb commented:
/lgtm

childsb commented:
/lgtm

gnufied commented:
/retest

humblec commented:
/retest

gnufied commented:
Appears to be flaking on - https://github.com/kubernetes/kubernetes/issues/49613 and some kubemark tests. The failures look unrelated to the PR.

gnufied commented:
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-e2e-kops-aws

jsafrane commented:
/retest

gnufied commented:
@childsb @jsafrane can you please readd lgtm. I just rebased!

gnufied commented:
/lgtm

childsb commented:
/retest

gnufied committed:
Add rbac policy change for expand controller

gnufied committed:
Implement controller for resizing volumes

gnufied committed:
Implement necessary API changes

Introduce feature gate for expanding PVs
Add a field to SC
Add new Conditions and feature tag pvc update
Add tests for size update via feature gate
register the resize admission plugin
Update golint failures

gnufied committed:
Update generated files - api, bazel, json

Update all generated files json, protocolbuffers,
documentation

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/lgtm

verult commented on pkg/volume/plugins.go:
```diff
@@ -201,6 +202,14 @@ type AttachableVolumePlugin interface {
+// expanded
+type ExpandableVolumePlugin interface {
+	VolumePlugin
+	ExpandVolumeDevice(spec *Spec, newSize resource.Quantity, oldSize resource.Quantity) (resource.Quantity, error)
```
Does it return the new size of the volume after the operation? These methods should have comments.

verult replied:
What's the unit for these sizes?

gnufied replied:
Yes it does return new size after volume operation.  I agree about comments. Added a github issue for myself so as I don't foget - https://github.com/kubernetes/kubernetes/issues/56177
